[shared-data] create table failed with hdfs storage volume

【详述】把2x集群的表，在3x（hfds存算分离）重建一遍，开始是OK的，建了几个小时后，fe报错Caused by: com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: Unexpected exception: Failed to get primary backend. shard id: 344953
其中一个cn，cpu打高一直gc。

【是否存算分离】是
【StarRocks版本】例如：3.2.12
【集群规模】例如：1fe +3cn
【机器信息】16c 60g
cpu高cn的stack，jmap统计，线程top cpu信息见：
https://forum.mirrorship.cn/t/topic/17333

--- a/be/src/common/config.h
+++ b/be/src/common/config.h
@@ -1047,6 +1047,7 @@ CONF_mInt32(starlet_fslib_s3client_connect_timeout_ms, "1000");
 // NOTE: need to handle the negative value properly
 CONF_Alias(object_storage_request_timeout_ms, starlet_fslib_s3client_request_timeout_ms);
 CONF_mInt32(starlet_delete_files_max_key_in_batch, "1000");
+CONF_mInt32(starlet_filesystem_instance_cache_capacity, "10000");
 #endif
 
 CONF_mInt64(lake_metadata_cache_limit, /*2GB=*/"2147483648");

--- a/be/src/http/action/update_config_action.cpp
+++ b/be/src/http/action/update_config_action.cpp
@@ -368,6 +368,15 @@ Status UpdateConfigAction::update_config(const std::string& name, const std::str
         UPDATE_STARLET_CONFIG(s3_use_list_objects_v1, fslib_s3client_use_list_objects_v1);
         UPDATE_STARLET_CONFIG(starlet_delete_files_max_key_in_batch, delete_files_max_key_in_batch);
 #undef UPDATE_STARLET_CONFIG
+
+        _config_callback.emplace("starlet_filesystem_instance_cache_capacity", [&]() -> Status {
+            LOG(INFO) << "set starlet_filesystem_instance_cache_capacity:"
+                      << config::starlet_filesystem_instance_cache_capacity;
+            if (g_worker) {
+                g_worker->set_fs_cache_capacity(config::starlet_filesystem_instance_cache_capacity);
+            }
+            return Status::OK();
+        });
 #endif // USE_STAROS
     });
 

--- a/be/src/service/staros_worker.cpp
+++ b/be/src/service/staros_worker.cpp
@@ -77,11 +77,17 @@ std::unique_ptr<staros::starlet::Starlet> g_starlet;
 
 namespace fslib = staros::starlet::fslib;
 
-StarOSWorker::StarOSWorker() : _mtx(), _shards(), _fs_cache(new_lru_cache(1024)) {}
+StarOSWorker::StarOSWorker()
+        : _mtx(), _shards(), _fs_cache(new_lru_cache(config::starlet_filesystem_instance_cache_capacity)) {}
 
 StarOSWorker::~StarOSWorker() = default;
 
 static const uint64_t kUnknownTableId = UINT64_MAX;
+
+void StarOSWorker::set_fs_cache_capacity(int32_t capacity) {
+    _fs_cache->set_capacity(capacity);
+}
+
 uint64_t StarOSWorker::get_table_id(const ShardInfo& shard) {
     const auto& properties = shard.properties;
     auto iter = properties.find("tableId");
@@ -135,9 +141,7 @@ absl::Status StarOSWorker::invalidate_fs(const ShardInfo& info) {
     if (!conf.ok()) {
         return conf.status();
     }
-    std::string key_str = get_cache_key(*scheme, *conf);
-    CacheKey key(key_str);
-    _fs_cache->erase(key);
+    erase_fs_cache(get_cache_key(*scheme, *conf));
     return absl::OkStatus();
 }
 
@@ -214,8 +218,10 @@ absl::StatusOr<std::shared_ptr<fslib::FileSystem>> StarOSWorker::get_shard_files
             l.unlock();
             return build_filesystem_on_demand(id, conf);
         }
-        if (it->second.fs) {
-            return it->second.fs;
+
+        auto fs = lookup_fs_cache(it->second.fs_cache_key);
+        if (fs != nullptr) {
+            return fs;
         }
     }
 
@@ -228,15 +234,18 @@ absl::StatusOr<std::shared_ptr<fslib::FileSystem>> StarOSWorker::get_shard_files
             l.unlock();
             return build_filesystem_on_demand(id, conf);
         }
-        if (shard_iter->second.fs) {
-            return shard_iter->second.fs;
+
+        auto fs = lookup_fs_cache(shard_iter->second.fs_cache_key);
+        if (fs != nullptr) {
+            return fs;
         }
+
         auto fs_or = build_filesystem_from_shard_info(shard_iter->second.shard_info, conf);
         if (!fs_or.ok()) {
             return fs_or.status();
         }
-        shard_iter->second.fs = std::move(fs_or).value();
-        return shard_iter->second.fs;
+        shard_iter->second.fs_cache_key = std::move(fs_or->first);
+        return fs_or->second;
     }
 }
 
@@ -260,11 +269,17 @@ absl::StatusOr<std::shared_ptr<fslib::FileSystem>> StarOSWorker::build_filesyste
     if (!info_or.ok()) {
         return info_or.status();
     }
-    return build_filesystem_from_shard_info(info_or.value(), conf);
+    auto fs_or = build_filesystem_from_shard_info(info_or.value(), conf);
+    if (!fs_or.ok()) {
+        return fs_or.status();
+    }
+
+    // Do not return the cache key shared_ptr, so if it not held by anyone else, the fs instance will be removed from the fs cache immediately.
+    return fs_or->second;
 }
 
-absl::StatusOr<std::shared_ptr<fslib::FileSystem>> StarOSWorker::build_filesystem_from_shard_info(
-        const ShardInfo& info, const Configuration& conf) {
+absl::StatusOr<std::pair<std::shared_ptr<std::string>, std::shared_ptr<fslib::FileSystem>>>
+StarOSWorker::build_filesystem_from_shard_info(const ShardInfo& info, const Configuration& conf) {
     auto localconf = build_conf_from_shard_info(info);
     if (!localconf.ok()) {
         return localconf.status();
@@ -316,23 +331,17 @@ absl::StatusOr<fslib::Configuration> StarOSWorker::build_conf_from_shard_info(co
     return info.fslib_conf_from_this(need_enable_cache(info), "");
 }
 
-absl::StatusOr<std::shared_ptr<StarOSWorker::FileSystem>> StarOSWorker::new_shared_filesystem(
-        std::string_view scheme, const Configuration& conf) {
-    std::string key_str = get_cache_key(scheme, conf);
-    CacheKey key(key_str);
+absl::StatusOr<std::pair<std::shared_ptr<std::string>, std::shared_ptr<fslib::FileSystem>>>
+StarOSWorker::new_shared_filesystem(std::string_view scheme, const Configuration& conf) {
+    std::string cache_key = get_cache_key(scheme, conf);
 
     // Lookup LRU cache
-    std::shared_ptr<fslib::FileSystem> fs;
-    auto handle = _fs_cache->lookup(key);
-    if (handle != nullptr) {
-        auto value = static_cast<CacheValue*>(_fs_cache->value(handle));
-        fs = value->lock();
-        _fs_cache->release(handle);
-        if (fs != nullptr) {
-            VLOG(9) << "Share filesystem";
-            return std::move(fs);
-        }
+    auto value_or = find_fs_cache(cache_key);
+    if (value_or.ok()) {
+        VLOG(9) << "Share filesystem";
+        return value_or;
     }
+
     VLOG(9) << "Create a new filesystem";
 
     // Create a new instance of FileSystem
@@ -341,18 +350,12 @@ absl::StatusOr<std::shared_ptr<StarOSWorker::FileSystem>> StarOSWorker::new_shar
         return fs_or.status();
     }
     // turn unique_ptr to shared_ptr
-    fs = std::move(fs_or).value();
+    std::shared_ptr<fslib::FileSystem> fs = std::move(fs_or).value();
 
     // Put the FileSysatem into LRU cache
-    auto value = new CacheValue(fs);
-    handle = _fs_cache->insert(key, value, 1, 1, cache_value_deleter);
-    if (handle == nullptr) {
-        delete value;
-    } else {
-        _fs_cache->release(handle);
-    }
+    auto fs_cache_key = insert_fs_cache(cache_key, fs);
 
-    return std::move(fs);
+    return std::make_pair(std::move(fs_cache_key), std::move(fs));
 }
 
 std::string StarOSWorker::get_cache_key(std::string_view scheme, const Configuration& conf) {
@@ -365,6 +368,73 @@ std::string StarOSWorker::get_cache_key(std::string_view scheme, const Configura
     }
     sha256.digest();
     return sha256.hex();
+}
+
+std::shared_ptr<std::string> StarOSWorker::insert_fs_cache(const std::string& key,
+                                                           const std::shared_ptr<FileSystem>& fs) {
+    std::shared_ptr<std::string> fs_cache_key(new std::string(key), [](std::string* key) {
+        if (g_worker) {
+            g_worker->erase_fs_cache(*key);
+        }
+        delete key;
+    });
+
+    CacheKey cache_key(key);
+    auto value = new CacheValue(fs_cache_key, fs);
+    auto handle = _fs_cache->insert(cache_key, value, 1, 1, cache_value_deleter);
+    if (handle == nullptr) {
+        delete value;
+        return nullptr;
+    }
+
+    _fs_cache->release(handle);
+    return fs_cache_key;
+}
+
+void StarOSWorker::erase_fs_cache(const std::string& key) {
+    CacheKey cache_key(key);
+    _fs_cache->erase(key);
+}
+
+std::shared_ptr<fslib::FileSystem> StarOSWorker::lookup_fs_cache(const std::shared_ptr<std::string>& key) {
+    if (key == nullptr) {
+        return nullptr;
+    }
+    return lookup_fs_cache(*key);
+}
+
+std::shared_ptr<fslib::FileSystem> StarOSWorker::lookup_fs_cache(const std::string& key) {
+    auto value_or = find_fs_cache(key);
+    if (!value_or.ok()) {
+        return nullptr;
+    }
+
+    return value_or->second;
+}
+
+absl::StatusOr<std::pair<std::shared_ptr<std::string>, std::shared_ptr<fslib::FileSystem>>> StarOSWorker::find_fs_cache(
+        const std::string& key) {
+    if (key.empty()) {
+        return absl::InvalidArgumentError("key is empty");
+    }
+
+    CacheKey cache_key(key);
+    auto handle = _fs_cache->lookup(cache_key);
+    if (handle == nullptr) {
+        return absl::NotFoundError(key + " not found");
+    }
+
+    auto value = static_cast<CacheValue*>(_fs_cache->value(handle));
+    // The value->key may be expired in a very short critical moment.
+    // At that moment, the value->key is not referenced by anyone but it's shared_ptr deleter haven't be executed,
+    // so the item haven't be removed from cache yet.
+    // In this situation, this function will return a null key and a valid fs instance.
+    // So the caller cannot assume the returned key always valid.
+    auto ret = std::make_pair(value->key.lock(), value->fs);
+
+    _fs_cache->release(handle);
+
+    return ret;
 }
 
 Status to_status(const absl::Status& absl_status) {

--- a/be/src/service/staros_worker.h
+++ b/be/src/service/staros_worker.h
@@ -77,15 +77,22 @@ public:
     // register the listener(callback) when new shard is added to the worker
     void register_add_shard_listener(add_shard_listener listener) { _add_shard_listener = std::move(listener); }
 
+    void set_fs_cache_capacity(int32_t capacity);
+
 private:
     struct ShardInfoDetails {
         ShardInfo shard_info;
-        std::shared_ptr<FileSystem> fs;
+        std::shared_ptr<std::string> fs_cache_key;
 
         ShardInfoDetails(const ShardInfo& info) : shard_info(info) {}
     };
 
-    using CacheValue = std::weak_ptr<FileSystem>;
+    struct CacheValue {
+        std::weak_ptr<std::string> key;
+        std::shared_ptr<FileSystem> fs;
+
+        CacheValue(const std::weak_ptr<std::string>& key, const std::shared_ptr<FileSystem>& fs) : key(key), fs(fs) {}
+    };
 
     // This function can be made static perfectly. The only reason to make it `virtual`
     // is, for unit test MOCK as it is the only interface to interact with g_starlet.
@@ -110,12 +117,20 @@ private:
     uint64_t get_table_id(const ShardInfo& shared_info);
 
     absl::StatusOr<std::shared_ptr<FileSystem>> build_filesystem_on_demand(ShardId id, const Configuration& conf);
-    absl::StatusOr<std::shared_ptr<FileSystem>> build_filesystem_from_shard_info(const ShardInfo& info,
-                                                                                 const Configuration& conf);
-    absl::StatusOr<std::shared_ptr<FileSystem>> new_shared_filesystem(std::string_view scheme,
-                                                                      const Configuration& conf);
+    absl::StatusOr<std::pair<std::shared_ptr<std::string>, std::shared_ptr<FileSystem>>>
+    build_filesystem_from_shard_info(const ShardInfo& info, const Configuration& conf);
+    absl::StatusOr<std::pair<std::shared_ptr<std::string>, std::shared_ptr<FileSystem>>> new_shared_filesystem(
+            std::string_view scheme, const Configuration& conf);
     absl::Status invalidate_fs(const ShardInfo& shard);
 
+    std::shared_ptr<std::string> insert_fs_cache(const std::string& key, const std::shared_ptr<FileSystem>& fs);
+    void erase_fs_cache(const std::string& key);
+    std::shared_ptr<FileSystem> lookup_fs_cache(const std::string& key);
+    std::shared_ptr<FileSystem> lookup_fs_cache(const std::shared_ptr<std::string>& key);
+    absl::StatusOr<std::pair<std::shared_ptr<std::string>, std::shared_ptr<FileSystem>>> find_fs_cache(
+            const std::string& key);
+
+private:
     mutable std::shared_mutex _mtx;
     std::unordered_map<ShardId, ShardInfoDetails> _shards;
     std::unique_ptr<Cache> _fs_cache;

--- a/be/test/service/staros_worker_test.cpp
+++ b/be/test/service/staros_worker_test.cpp
@@ -15,9 +15,13 @@
 #ifdef USE_STAROS
 #include "service/staros_worker.h"
 
+#include <fslib/configuration.h>
+#include <fslib/fslib_all_initializer.h>
 #include <gtest/gtest.h>
 
 #include <functional>
+
+#include "common/config.h"
 
 namespace starrocks {
 
@@ -60,6 +64,53 @@ TEST(StarOSWorkerTest, test_add_listener) {
     // no change, the shard:2 is already added
     EXPECT_EQ(1, counter);
     EXPECT_EQ(1, ids.size());
+}
+
+TEST(StarOSWorkerTest, test_fs_cache) {
+    staros::starlet::fslib::register_builtin_filesystems();
+    staros::starlet::ShardInfo shard_info;
+    shard_info.id = 1;
+    auto fs_info = shard_info.path_info.mutable_fs_info();
+    fs_info->set_fs_type(staros::FileStoreType::S3);
+    auto s3_fs_info = fs_info->mutable_s3_fs_info();
+    s3_fs_info->set_bucket("test_bucket");
+    s3_fs_info->set_endpoint("test_endpoint");
+    s3_fs_info->set_region("us-east-1");
+    auto credential = s3_fs_info->mutable_credential();
+    auto simple_credential = credential->mutable_simple_credential();
+    simple_credential->set_access_key("test_ak");
+    simple_credential->set_access_key_secret("test_sk");
+    // set full path
+    shard_info.path_info.set_full_path(absl::StrFormat("s3://%s/%d/", s3_fs_info->bucket(), time(NULL)));
+
+    // cache settings
+    shard_info.cache_info.set_enable_cache(false);
+    shard_info.cache_info.set_async_write_back(false);
+
+    auto schema_or = StarOSWorker::build_scheme_from_shard_info(shard_info);
+    EXPECT_TRUE(schema_or.ok());
+    auto schema = schema_or.value();
+
+    auto conf_or = shard_info.fslib_conf_from_this(false, "");
+    EXPECT_TRUE(conf_or.ok());
+    auto conf = conf_or.value();
+
+    auto cache_key = StarOSWorker::get_cache_key(schema, conf);
+
+    auto worker = std::make_shared<StarOSWorker>();
+    g_worker = worker;
+
+    EXPECT_TRUE(worker->add_shard(shard_info).ok());
+
+    EXPECT_FALSE(worker->lookup_fs_cache(cache_key));
+
+    EXPECT_TRUE(worker->get_shard_filesystem(shard_info.id, conf).ok());
+
+    EXPECT_TRUE(worker->lookup_fs_cache(cache_key));
+
+    EXPECT_TRUE(worker->remove_shard(shard_info.id).ok());
+
+    EXPECT_FALSE(worker->lookup_fs_cache(cache_key));
 }
 
 } // namespace starrocks

